File: convert.py
import os
import json
import shutil
from PIL import Image

# Directory containing the files
base_directory = "images"

# Path to the JSON file
json_path = "data.json"

# Convert images to WebP
def convert_images_to_webp(base_directory):
    for root, _, files in os.walk(base_directory):
        if root.endswith("backup") or "/backup/" in root or "\\backup\\" in root:
            continue

        backup_folder = os.path.join(root, "backup")
        os.makedirs(backup_folder, exist_ok=True)

        for file in files:
            if file.lower().endswith((".jpg", ".png")):
                filepath = os.path.join(root, file)
                output_path = os.path.splitext(filepath)[0] + ".webp"

                with Image.open(filepath) as img:
                    img.save(output_path, "webp")
                
                print(f"Converted: {filepath} -> {output_path}")

                backup_path = os.path.join(backup_folder, file)
                shutil.move(filepath, backup_path)
                print(f"Moved: {filepath} -> {backup_path}")

# Update the JSON file
def update_json_file(json_path):
    with open(json_path, "r", encoding="utf-8") as f:
        data = json.load(f)
    
    def update_image_paths(obj):
        if isinstance(obj, dict):
            for key, value in obj.items():
                if key == "images" and isinstance(value, list):
                    obj[key] = [
                        img.replace(".jpg", ".webp").replace(".png", ".webp") for img in value
                    ]
                else:
                    update_image_paths(value)
        elif isinstance(obj, list):
            for item in obj:
                update_image_paths(item)

    update_image_paths(data)

    with open(json_path, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=4, ensure_ascii=False)

    print("JSON updated!")

# Execute the steps
convert_images_to_webp(base_directory)
update_json_file(json_path)

File: convert.js
import fs from "node:fs";
import path from "node:path";
import sharp from "sharp";

const BASE_DIR = "src/images";
const JSON_PATH = "src/data.json";

async function convertImagesToWebp(directory) {
	const entries = fs.readdirSync(directory, { withFileTypes: true });

	const backupDir = path.join(directory, "backup");
	if (!fs.existsSync(backupDir)) {
	}

	for (const entry of entries) {
		const fullPath = path.join(directory, entry.name);

		if (entry.isDirectory()) {
			if (entry.name !== "backup") {
				await convertImagesToWebp(fullPath);
			}
			continue;
		}

		const ext = path.extname(entry.name).toLowerCase();
		if (ext === ".jpg" || ext === ".png") {
			fs.mkdirSync(backupDir, { recursive: true });

			const outputPath = path.join(
				directory,
				`${path.parse(entry.name).name}.webp`,
			);
			const backupPath = path.join(backupDir, entry.name);

			try {
				await sharp(fullPath).toFile(outputPath);
				console.log(`Converted: ${fullPath} -> ${outputPath}`);

				fs.renameSync(fullPath, backupPath);
				console.log(`Moved: ${fullPath} -> ${backupPath}`);
			} catch (err) {
				console.error(`Error processing ${fullPath}:`, err);
			}
		}
	}
}

function updateJsonFile(filePath) {
	const rawData = fs.readFileSync(filePath, "utf8");
	const data = JSON.parse(rawData);
	let modified = false;

	function updateImagePaths(obj) {
		if (Array.isArray(obj)) {
			for (const item of obj) {
				updateImagePaths(item);
			}
		} else if (typeof obj === "object" && obj !== null) {
			for (const key in obj) {
				if (key === "images" && Array.isArray(obj[key])) {
					const newImages = obj[key].map((img) => {
						const newImg = img.replace(/\.jpg$|\.png$/i, ".webp");
						if (newImg !== img) modified = true;
						return newImg;
					});
					obj[key] = newImages;
				} else {
					updateImagePaths(obj[key]);
				}
			}
		}
	}

	updateImagePaths(data);

	if (modified) {
		fs.writeFileSync(filePath, JSON.stringify(data, null, 4), "utf8");
		console.log("JSON updated!");
	} else {
		console.log("JSON already up to date.");
	}
}

(async () => {
	console.log("Starting conversion...");
	await convertImagesToWebp(BASE_DIR);
	updateJsonFile(JSON_PATH);
	console.log("Done.");
})();
